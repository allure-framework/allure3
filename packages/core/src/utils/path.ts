import { pathToFileURL } from "node:url";

/**
 * The function appends `file:///` protocol to the absolute paths on Windows due to ES modules imports specifics
 * @param path
 */
export const normalizeImportPath = (path: string) => {
  return pathToFileURL(path).href;
};
