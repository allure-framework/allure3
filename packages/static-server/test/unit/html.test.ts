import { layer } from "allure-js-commons";
import axios, { type AxiosError } from "axios";
import getPort from "get-port";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { beforeEach, describe, expect, it } from "vitest";
import { type AllureStaticServer, serve } from "../../src/index.js";

const baseDir = dirname(fileURLToPath(import.meta.url));
const servePath = join(baseDir, "fixtures");

let server: AllureStaticServer;
let port: number;

beforeEach(async () => {
  await layer("unit");

  port = await getPort();
  server?.stop();
});

it("returns code 404 when the requested file doesn't exist", async () => {
  server = await serve({ port, servePath });

  try {
    await axios.get(`http://localhost:${port}/404.html`, {
      timeout: 500,
    });
  } catch (err) {
    expect((err as AxiosError).response).toMatchObject({
      status: 404,
    });
  }
});

it("returns index.html file by the full path", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}/index.html`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns index.html when an url ends with /, when the file exists", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}/`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns page with available files when an url ends with / and index.html doesn't exist", async () => {
  server = await serve({ port, servePath: join(servePath, "withoutIndex") });
  const res = await axios.get(`http://localhost:${port}/`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns page with available files when an url ends with / asd", async () => {
  server = await serve({ port, servePath: join(servePath, "withoutIndex", "baz") });
  const res = await axios.get(`http://localhost:${port}/`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns index.html when an url ends just with word without file extension", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns nested/index.html file by the full path", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}/nested/index.html`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns nested/index.html when an url ends with /", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}/nested/`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

it("returns nested/index.html when an url ends just with word without file extension", async () => {
  server = await serve({ port, servePath });
  const res = await axios.get(`http://localhost:${port}/nested`, {
    timeout: 500,
    headers: {
      "Content-Type": "text/html",
    },
  });

  expect(res.status).toBe(200);
  expect(res.data).toMatchSnapshot();
});

describe("live", () => {
  it("injects live reload script", async () => {
    server = await serve({ port, servePath, live: true });
    const res = await axios.get(`http://localhost:${port}/index.html`, {
      timeout: 500,
      headers: {
        "Content-Type": "text/html",
      },
    });

    expect(res.status).toBe(200);
    expect(res.data).toMatchSnapshot();
  });

  it("doesn't inject live reload script to html with attachment query parameter", async () => {
    server = await serve({ port, servePath, live: true });
    const res = await axios.get(`http://localhost:${port}/index.html?attachment`, {
      timeout: 500,
      headers: {
        "Content-Type": "text/html",
      },
    });

    expect(res.status).toBe(200);
    expect(res.data).toMatchSnapshot();
  });
});
