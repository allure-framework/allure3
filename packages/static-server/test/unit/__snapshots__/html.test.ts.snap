// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`live > doesn't inject live reload script to html with attachment query parameter 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample page</title>
</head>
<body>
<h1>Hello world!</h1>
</body>
</html>"
`;

exports[`live > injects live reload script 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns index.html file by the full path 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns index.html when an url ends just with word without file extension 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns index.html when an url ends with /, when the file exists 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns nested/index.html file by the full path 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample nested page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns nested/index.html when an url ends just with word without file extension 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample nested page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns nested/index.html when an url ends with / 1`] = `
"<!doctype html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <title>Sample nested page</title>
</head>
<body>
<h1>Hello world!</h1>

    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
</html>"
`;

exports[`returns page with available files when an url ends with / and index.html doesn't exist 1`] = `
"
    <!DOCTYPE html>
    <html lang="en">
      <head>
        <title>Allure Server</title>
        <meta charset="UTF-8" />
      </head>
      <body>
        <p>Allure Static Server</p>
        <ul>
          
          <li><a href="./bar/">bar/</a></li><li><a href="./baz/">baz/</a></li><li><a href="./foo/">foo/</a></li><li><a href="./foo.html">foo.html</a></li>
        </ul>
      
    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
    </html>
  "
`;

exports[`returns page with available files when an url ends with / asd 1`] = `
"
    <!DOCTYPE html>
    <html lang="en">
      <head>
        <title>Allure Server</title>
        <meta charset="UTF-8" />
      </head>
      <body>
        <p>Allure Static Server</p>
        <ul>
          
          <li><a href="./foo/">foo/</a></li><li><a href="./foo.html">foo.html</a></li>
        </ul>
      
    <script>
      const eventSource = new EventSource("/__live_reload");
      
      eventSource.onmessage = () => {
        const newHash = Math.random().toString(36).slice(2, 10);
        
        window.localStorage.setItem("__allure_report_live_reload_hash__", newHash);
        window.location.reload();
      };
    </script>
  </body>
    </html>
  "
`;
